### PR TITLE
oauth2/api: pass object instead of list of params

### PR DESCRIPTION
**Pull request by VasilMitovClouway**
In Tokens now an object TokenRequest is passed instead of parameters for better readability. 
A builder has been also added.

I'm not sure i should keep these long tests for equality for TokenRequest.

Fixes: #73

### CODE REVIEW

File: oauth2-server/src/main/java/com/clouway/oauth2/token/TokenRequest.java [hunk 1]
```diff
@@ -1,44 +1,110 @@
...
-  public final String clientSecret;
+  public final GrantType grantType;
+  public final Client client;
+  public final Identity identity;
+  public final Set<String> scopes;
+  public final DateTime when;
+  public final Map<String, String> params;
 
-  public TokenRequest(String grantType, String code, String refreshToken, String clientId, String clientSecret) {
+  public TokenRequest(GrantType grantType, Client client, Identity identity, Set<String> scopes, DateTime when, Map<String, String> params) {
```
Comment (mgenov) [inline oauth2-server/src/main/java/com/clouway/oauth2/token/TokenRequest.java]: This constructor should be removed too and only builder usage to be enforced. 

Passing of many params as method arguments is similar as passing of many params to the constructor.

File: oauth2-server/src/main/java/com/clouway/oauth2/token/TokenRequest.java [hunk 2]
```diff
@@ -1,44 +1,110 @@
...
-    int result = grantType != null ? grantType.hashCode() : 0;
-    result = 31 * result + (code != null ? code.hashCode() : 0);
-    result = 31 * result + (clientId != null ? clientId.hashCode() : 0);
-    result = 31 * result + (clientSecret != null ? clientSecret.hashCode() : 0);
-    return result;
+    return Objects.hashCode(grantType, client, identity, scopes, when, params);
+  }
+
+
+  public static final class Builder {
```
Comment (mgenov) [inline oauth2-server/src/main/java/com/clouway/oauth2/token/TokenRequest.java]: Inner classes should be defined first.

File: oauth2-server/src/test/java/com/clouway/oauth2/token/TokenRequestEqualityTest.java [hunk 3]
```diff
@@ -12,17 +20,164 @@
 public class TokenRequestEqualityTest {
   @Test
   public void areEqual() {
-    TokenRequest request1 = new TokenRequest("grant_type", "code","refresh_token", "client_id", "client_secret");
-    TokenRequest request2 = new TokenRequest("grant_type", "code","refresh_token", "client_id", "client_secret");
+    DateTime anyDateTime = new DateTime();
```
Comment (mgenov) [inline oauth2-server/src/test/java/com/clouway/oauth2/token/TokenRequestEqualityTest.java]: This test looks a bit overwhelming with few instantiations of whole object. You can simplify it by using 2 cases: equal and not equal. 

EqualsVerifier looks like a library which could be tested about this: http://jqno.nl/equalsverifier/manual/getting-started/

**Review by mgenov — CHANGES_REQUESTED** [review]

File: oauth2-server/src/test/java/com/clouway/oauth2/token/TokenRequestEqualityTest.java (hunk 3 above)
Comment (VasilMitovClouway) [reply oauth2-server/src/test/java/com/clouway/oauth2/token/TokenRequestEqualityTest.java]: With the implementation of this library should i change the Equals tests for other classes as well or should i do it in a separate PR?

File: oauth2-server/src/main/java/com/clouway/oauth2/token/TokenRequest.java (hunk 1 above)
Comment (VasilMitovClouway) [reply oauth2-server/src/main/java/com/clouway/oauth2/token/TokenRequest.java]: Removed.

File: oauth2-server/src/main/java/com/clouway/oauth2/token/TokenRequest.java (hunk 2 above)
Comment (VasilMitovClouway) [reply oauth2-server/src/main/java/com/clouway/oauth2/token/TokenRequest.java]: Moved to be first.

File: oauth2-server/src/test/java/com/clouway/oauth2/token/TokenRequestEqualityTest.java (hunk 3 above)
Comment (mgenov) [reply oauth2-server/src/test/java/com/clouway/oauth2/token/TokenRequestEqualityTest.java]: In separate PR sounds better.

**Review by mgenov — APPROVED** [review]
LGTM !

**Pull request merged**